### DISCUSSION TITLE
Can the XMLRPC server listen on all (or a configurable) IP address ?

### DISCUSSION THREAD

**Discussion by thiloplanz** [SukramJ/aiohomematic#1722] [Q&A] [answered]
Hi,

I am running HA in Docker (using Orbstack on a Mac, with host networking mode), and after an update the XMLRPC Pong stopped working.

I believe this is caused by a recent change that made the XMLRPC server no longer listen on all IP addresses, but only on an auto-detected local one. https://github.com/danielperna84/hahomematic/pull/1655
As a result, I can only access the XMLRPC port from the local machine (both docker containers and from the Mac host), but no longer from other machines (because that auto-detected IP is not routed over the network), in particular not from the CCU.

Is it possible to configure this listening address?

In the mean-time I'll try to make it work with some proxy server.

**Comment by SukramJ** [comment]
The context is missing here completely without any configurations, diagnostics and network diagrams!
Did you configure the callback host/port?
How did you configure your HA docker network (host only)?
Did you read the integration's documentation and searched for docker?

> As a result, I can only access the XMLRPC port from the local machine (both docker containers and from the Mac host), but no longer from other machines

Why do you want to access the port from a machine other then your CCU?

**Reply by thiloplanz** [reply]
This used to work with an older version, and it still works with a proxy.

> Did you configure the callback host/port?

I did. But the host I put in there does not get picked up by the code that starts the server anymore (it does use the port still).

> Why do you want to access the port from a machine other then your CCU?

The CCU is running on a different machine than Home Assistant, so it will need to call back into that XMLRPC on the HA via the network.
I can see the log files saying that HA is setting up a listening socket.

```
2024-09-26 17:25:12.099 DEBUG (XmlRpcServer 198.19.249.2:20123) [hahomematic.central.xml_rpc_server] RUN: Starting XmlRPC-Server at http://198.19.249.2:20123
2024-09-26 17:25:12.099 DEBUG (MainThread) [hahomematic.central.xml_rpc_server] CREATE_XML_RPC_SERVER: Starting XmlRPC-Server on 198.19.249.2:20123
```

Unfortunately the IP address it uses there is not externally accessible. In an older version, I think it was listening on all IP (0.0.0.0), or maybe to the one configured in the Callback Hostname/IP-Address configuration (which the CCU will be using to try to reach HA).

**Reply by thiloplanz** [reply]
I have reverted back to an older version of the integration (1.60.0) and the Ping/Pong works again.
It did indeed listen on all available IP addresses, including one that is accessible over the bridge network (the one I configured in 
the Callback Hostname/IP-Address box, the one that is communicated to CCU).

```
2024-09-26 18:02:50.153 DEBUG (XmlRpcServer on port 20123) [hahomematic.central.xml_rpc_server] RUN: Starting XmlRPC-Server at http://0.0.0.0:20123
2024-09-26 18:02:50.153 DEBUG (MainThread) [hahomematic.central.xml_rpc_server] REGISTER_XML_RPC_SERVER: Starting XmlRPC-Server
```

**Reply by SukramJ** [reply]
> The context is missing here completely without any configurations, diagnostics and network diagrams!

Mit Beschreibungen kann ich nichts anfangen, wenn keine Fakten drin sind.
Welche IP hat HA /der HA Host?
Welche IP hat die CCU?
Woher kommt die 198.19.249.2?

Häng bitte die Integrationsdiagnose, die HA-DockerKonfiguration an und andere Konfigurationen und Diagramme die hilfreich zum Verständnis sein könnten.

-

**Reply by thiloplanz** [reply]
Aus dem verlinkten Merge-Request ist das eigentlich recht ersichtlich.

In der alten Version von `create_xml_rpc_server` wird keine IP-Adresse spezifiziert und der Server bindet sich an alle verfügbaren.

In der neuen Version wird die Funktion `get_ip_addr` benutzt, um eine einzelne IP-Adresse zu finden, die dann für den Server verwendet wird. In meinem Fall ist das die lokale IP das virtuellen Linux-Host auf dem Docker läuft. Interessanterweise bekommt `get_ip_addr` allerdings den konfigurierten Callback-Host als Parameter, vielleicht hätte das dann doch irgendwie klappen können? 

https://github.com/danielperna84/hahomematic/pull/1655/files#diff-ca173bbd8dbe379b44fd70743e4b6971960512bc6e6866a41865c2679a513ae7R255

In meinem Fall gibt es die folgenden IP:

198.19.249.2   Die IP den Dockerhosts (virtuelle Machine mit Linux), der Container mit HA hat die selbe IP, da Networkmodus "Host". `get_ip_addr` findet diese.

192.168.2.97 Die Netzwerkadresses des Mac, auf dem HA läuft. Diese Adresse kann die CCU aufrufen, und das ist auch als Callbackhost eingetragen. Dank Networkmodus Host biegt Orbstack das dann so hin, dass Pakete an 192.168.2.97 auch in dem HA-Container landen, aber der müsste dann auch auf dieser IP horchen

Sorry, muss Schluss machen für heute, ich hoffe die Beschreibung hilft ein wenig, es läuft jetzt erstmal alles in der alten Version wieder, ich fasse das besser erst wieder an, wenn die Heizsaison vorbei ist....

**Reply by SukramJ** [reply]
> Aus dem verlinkten Merge-Request ist das eigentlich recht ersichtlich.

Danke, ich kenn meinen Code und weiss was und warum ich das geändert habe.
Der Rest der Beschreibung hilft weiter, aber ich denke das ich vorerst nichts ändern werde.

**Reply by SukramJ** [reply]
@thiloplanz 
Bitte mal mit der Version 1.67.0b0 testen. Gegebenenfalls muss "Vorabversion/Pre-release" für die Integration in HACS aktiviert werden. 
Nach dem Neustart die Integration konfigurieren. 
Auf der zweiten Seite die erweiterte Konfiguration auswählen und auf der dritten Seite "Auf allen IP Adressen lauschen" aktivieren.

**Reply by thiloplanz** [reply]
Danke für den neuen Build.

Ich habe 1.67.0b0 aktiviert (glaube ich, es wird als 1.67.0 im Settings-UI angezeigt), aber auf der dritten Seite gibt es diese Option nicht. Kann ich die von Hand irgendwo reinschreiben?

![Bildschirmfoto 2024-10-08 um 18 40 47](https://github.com/user-attachments/assets/287d930c-b9cf-49b8-96a9-bb68ce638d1d)

Es gab auch noch 1.67.0b1, aber das will eine neuere dev-Version von HA.

**Reply by SukramJ** [reply, marked as answer]
Dein Browsercache ist veraltet! Refresh, und danach sind die Platzhalter weg.

1.67.0b1 ist aber nötig, und dazu HA 2024.10.x

**Reply by thiloplanz** [reply]
Jupp, nach dem Update auf HA 2024.10 und 1.67.0b1 erscheint die neue Option "Auf allen IP Adressen lauschen" , und wenn ich diese aktiviere funktioniert alles wieder.

💚
